JDK-8293114: JVM should trim the native heap

tstuefe opened the pull request:
This is a continuation of https://github.com/openjdk/jdk/pull/10085. I closed https://github.com/openjdk/jdk/pull/10085 because it had accumulated too much comment history and got confusing. For a history of this issue, see previous discussions [1] and the comment section of 10085.
 
---------------

This RFE adds the option to trim the Glibc heap periodically. This can recover a significant memory footprint if the VM process suffers from high-but-rare malloc spikes. It does not matter who causes the spikes: the JDK or customer code running in the JVM process.

### Background:

The Glibc is reluctant to return memory to the OS. Temporary malloc spikes often carry over as permanent RSS increase. Note that C-heap retention is difficult to observe. Since it is freed memory, it won't appear in NMT; it is just a part of RSS.

This is, effectively, caching - a performance tradeoff by the glibc. It makes a lot of sense with applications that cause high traffic on the C-heap. The JVM, however, clusters allocations and often rolls its own memory management based on virtual memory for many of its use cases.

To manually trim the C-heap, Glibc exposes `malloc_trim(3)`. With JDK 18 [2], we added a new jcmd command to *manually* trim the C-heap on Linux (`jcmd System.trim_native_heap`). We then observed customers running this command periodically to slim down process sizes of container-bound jvms. That is cumbersome, and the JVM can do this a lot better - among other things because it knows best when *not* to trim.

#### GLIBC internals

The following information I took from the glibc source code and experimenting.

##### Why do we need to trim manually? Does the Glibc not trim on free?

Upon `free()`, glibc may return memory to the OS if:
- the returned block was mmap'ed
- the returned block was not added to tcache or to fastbins
- the returned block, possibly merged with its two immediate neighbors, had they been free, is larger than FASTBIN_CONSOLIDATION_THRESHOLD (64K) - in that case:
  a) for the main arena, glibc attempts to lower the brk()
  b) for mmap-ed heaps, glibc attempts to completely unmap or shrink the heap.
In both cases, (a) and (b), only the top portion of the heap is reclaimed. "Holes" in the middle of other in-use chunks are not reclaimed.

So: glibc *may* automatically reclaim memory. In normal configurations, with typical C-heap allocation granularity, it is unlikely.

To increase the chance of auto-reclamation happening, one can do one or more things:
- a) increase allocation sizes
- b) limit mallocs to very few threads, ideally just one 
- c) set MALLOC_ARENA_MAX=1 
- d) set the `glibc.malloc.trim_threshold` to a very low value, e.g., 1

But:

- (a) is not possible for foreign code; not that even within hotspot, where we cluster allocations using hotspot arenas, the typical arena chunk size is too small to be auto-reclaimed
- (b) may just not be feasible
- (c) is *terrible* for performance since many C-Heap operations will compete over the lock of that single arena
- (d) works if either (b) or (c) are in place and if the released block happens to border the top of the arena. And it also costs performance.

The JVM only has limited influence on (a), none on (b), (c) is a really bad idea, and hence (d) often does little. That mirrors my practical experiences.

##### How does `malloc_trim()` differ from trimming on free() ?

`malloc_trim()`, will look for holes that are larger than a page; so it limits itself not to just reclaiming memory at the top of the arena. It will then `madvise(MADV_DONTNEED)` those holes. It does that for every arena.

##### What are the cons of calling `malloc_trim()`?

`malloc_trim()` cannot be interrupted. Once it runs, it runs. The runtime of `malloc_trim()` is not predictable. If there is nothing to reclaim, it is very fast (sub-ms). If there is a lot to reclaim (e.g. >32GB), I saw times of up to 800ms.

Moreover, `malloc_trim`, while trimming each arena, locks the arena. That may lock out concurrent C-heap operations in the thread that uses this arena. Note, however, that this is rare since many operations will be satisfied from the tcache and therefore don't lock.

##### What about the `pad` parameter for `malloc_trim()`

I found it has very little effect. It only affects how many bytes are preserved at the top of the main arena. It does not affect other arenas, nor does it affect how much space malloc_trim reclaims by releasing "holes", which is the main part of memory release.

### The Patch

Patch adds new options (experimental):

```
-XX:+GCTrimNativeHeap
-XX:GCTrimNativeHeapInterval=<seconds> (defaults to 60)
```

`GCTrimNativeHeap` is off by default. If enabled, it will cause the VM to trim the native heap periodically. The period is defined by `GCTrimNativeHeapInterval`.

Periodic trimming is done in its own thread. We cannot burden the ServiceThread, since the runtime of trims is unpredictable. 

The patch also adds a way to suspend trimming temporarily; if suspended, no trims will start, but ongoing trims will still finish.

The patch uses this mechanism to suspend trimming during GC STW phases and whenever we are about to do bulk C-heap operations (e.g. deleting deflated monitors).

### Examples:

This is an artificial test that causes two high malloc spikes with long idle periods. 

(yellow) NMT shows two spikes for malloc'ed memory;

(red) RSS of the baseline JVM shows that we reach a maximum and then never recover. This is the glibc retaining the free'd memory.

(blue) RSS of the patched JVM shows that we recover RSS in steps by doing periodic C-heap trimming.

![alloc-test](https://raw.githubusercontent.com/tstuefe/autotrim-experiments/master/alloc-c-heap-repro/basline-vs-autotrim30s.png)

(See here for parameters: [run script](http://cr.openjdk.java.net/~stuefe/other/autotrim/run-all.sh) )

### Tests

Tested older Glibc (2.31), and newer Glibc (2.35) (`mallinfo()` vs` mallinfo2()`), on Linux x64.

Older versions of this patch were routinely tested at SAP for almost half a year.

- [1] https://mail.openjdk.org/pipermail/hotspot-dev/2021-August/054323.html
- [2] https://bugs.openjdk.org/browse/JDK-8269345

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293114](https://bugs.openjdk.org/browse/JDK-8293114): JVM should trim the native heap (**Enhancement** - P4)


### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14781/head:pull/14781` \
`$ git checkout pull/14781`

Update a local copy of the PR: \
`$ git checkout pull/14781` \
`$ git pull https://git.openjdk.org/jdk.git pull/14781/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14781`

View PR using the GUI difftool: \
`$ git pr show -t 14781`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14781.diff">https://git.openjdk.org/jdk/pull/14781.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14781#issuecomment-1623399540)

jdksjolen commented:
>And app's malloc load can fluctuate wildly, with temporary spikes and long idle periods.

Are you talking about allocations into native memory that a Java application does on its own accord and not as a consequence of the JVM doing its own allocs? For compiling, for example.

tstuefe commented:
> > And app's malloc load can fluctuate wildly, with temporary spikes and long idle periods.
> 
> Are you talking about allocations into native memory that a Java application does on its own accord and not as a consequence of the JVM doing its own allocs? For compiling, for example.

This does not matter. The resulting malloc load is the sum of whatever we do and whatever the native code does.

tstuefe commented:
Ping @shipilev and @robehn

shipilev commented:
I don't think the comments from my yesterday's review were addressed :) There are some old comments (marked with `(Outdated)`, helpfully), but some are new. Please take a look at them?

tstuefe commented:
> I don't think the comments from my yesterday's review were addressed :) There are some old comments (marked with `(Outdated)`, helpfully), but some are new. Please take a look at them?

Of course, sorry. Seems I missed most of them.

About Pause in all VM ops, an alternative would be to just check in the trimmer if we are at safepoint, and if yes treat it as pause. I'll see if that's easier (I'm worried about pulling a mutex or atomic increasing the pauser variable in every VM op we run).

shipilev commented:
> About Pause in all VM ops, an alternative would be to just check in the trimmer if we are at safepoint, and if yes treat it as pause. I'll see if that's easier (I'm worried about pulling a mutex or atomic increasing the pauser variable in every VM op we run).

Oh yes, I like it. We can just check `SafepointSynchronize::is_at_safepoint()` and `SafepointSynchronize::is_synchronizing()`. Would be even better, because we could stop trimming when there is safepoint sync pending. Makes the whole thing much less intrusive.

jdksjolen commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -91,9 +91,26 @@ class ChunkPool {
+  }
+
+  static bool needs_cleaning() {
     for (int i = 0; i < _num_pools; i++) {
-      _pools[i].prune();
+      if (!_pools[i].empty()) {
+        return false;
+      }
+    }
+    return true;
+  }
```
Something seems wrong here? Having only empty pools means that `::prune()` is a no-op.

tstuefe replied:
Thanks for catching this.

tstuefe replied:
Somewhat scary that we don't have a test for that.

tstuefe commented:
@jdksjolen @shipilev New version:
- I removed all manual pause calls from all GC code and replaced those with just not trimming when at or near a safe point. This is less invasive since I expect the typical trim interval to be much larger than the interval we do our VM operations in.
- The pauses in runtime code remain
- I restored the original arena coding, I just added the pause. Though this coding could greatly benefit from cleanups, I want to keep this patch focused and easy to downport.
- Since we no longer have close ties to GC coding, the tests don't need to run per gc, so I dumbed down the tests.

@shipilev : I kept the SuspendMark inside TrimNative, because I like it that way. Otherwise, I would name it something like TrimNativeSuspendMark, so nothing gained but another symbol at global scope.

lkorinth commented on src/hotspot/share/runtime/trimNative.cpp:
```diff
@@ -0,0 +1,194 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questioSns.
```
This copyright notice differs in "questioSns" (extra 'S') and lack of  "DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER. ", might create problems for Oracle scripts

tstuefe replied:
Done. Thanks for catching that.

lkorinth commented:
The description says `-XX:GCTrimNativeHeapInterval=<seconds> (defaults to 60)`, but the code says milliseconds.

lkorinth commented on src/hotspot/share/runtime/trimNative.cpp:
```diff
@@ -0,0 +1,246 @@
+
+class NativeTrimmerThread : public NamedThread {
+
+  Monitor* const _lock;
```
Maybe inline this instead of having it a pointer? Even if the thread and lock is not destroyed until VM shutdown, I always feel the need to match new in constructors with delete in destructors.

tstuefe replied:
I rather avoid that. There is a timing factor here: JVM issues ThreadNative::cleanup, which stops the thread, but the thread is not guaranteed to stop before VM ends (may be in a middle of a trim operation). I don't want to have to think about this.

Ultimately, I don't see the point of the cleanup. I prefer a fast shutdown. Which is in line with a lot of code in the hotspot (e.g. we don't clean global mutexes either).

But I'll add an assert to make sure only one NativeTrimmerThread is ever created. So we know its a singleton.

lkorinth commented on src/hotspot/share/runtime/trimNative.cpp:
```diff
@@ -0,0 +1,246 @@
+  }
+  static constexpr int safepoint_poll_ms = 250;
+
+  static int64_t now() { return os::javaTimeMillis(); }
```
I think it would be better to not use CLOCK_REALTIME in case of clock changes by NTP etc.

tstuefe replied:
Okay, changed to os::elapsedTime().

shipilev commented on test/hotspot/jtreg/runtime/os/TestTrimNative.java:
```diff
@@ -0,0 +1,279 @@
+
+            System.out.println("Will spike now...");
+            for (int i = 0; i < numAllocations; i++) {
+                ptrs[i] = Unsafe.getUnsafe().allocateMemory(szAllocations);
```
Pull `Unsafe.getUnsafe()` into a local or a field?

shipilev commented on test/hotspot/jtreg/runtime/os/TestTrimNative.java:
```diff
@@ -0,0 +1,279 @@
+        } else if (args[0].equals("test")) {
+            runTest(Arrays.copyOfRange(args, 1, args.length));
+        } else if (args[0].equals("testOffOnNonCompliantPlatforms")) {
+            testOffOnNonCompliantPlatforms();
```
Inline these `test*`? This would give some symmetry against `runTest`.

shipilev commented on test/hotspot/jtreg/runtime/os/TestTrimNative.java:
```diff
@@ -0,0 +1,279 @@
+            throw new RuntimeException("Argument error");
+        }
+
+        if (args[0].equals("RUN")) {
```
The usual trick is to pull this "internal" main into a class, and then reference it. Like:

```
public Test {
 
  public void test() {
    runWith("...", "Test$Workload");
  }  

  public class Workload {
    public static void main(String... args) {} 
  }
}
```

shipilev commented on src/hotspot/share/runtime/trimNative.hpp:
```diff
@@ -0,0 +1,65 @@
+ *
+ */
+
+#ifndef SHARE_GC_SHARED_TRIMNATIVE_HPP
```
This is now `SHARE_RUNTIME_TRIMNATIVE_HPP`.

shipilev reviewed (commented):
Another read. I think we would need to tighten up style a bit too: the newline spacing style is different across the change.

robehn commented on src/hotspot/share/runtime/trimNative.cpp:
```diff
@@ -0,0 +1,246 @@
+        log_debug(trim)("Total trims: " UINT64_FORMAT ".", _num_trims_performed);
+        return true;
+      } else {
+        log_info(trim)("Trim native heap (no details)");
```
I would still like to know the trim_time.

tstuefe replied:
Done

robehn reviewed (approved):
Thanks, looks good to me.

dholmes-ora reviewed (commented):
I had an initial look at this. Seems okay in principle. The naming/terminology needs some updating IMO: "trimNative" doesn't convey enough information, please use "trimNativeHeap". "trim" for logging tag is also too non-descript.

As this is initially experimental I'm not overly concerned about the impact, though I do cringe at yet-another-VM-thread.

FYI I will be away until next Thursday, but no need to wait for me for further comments.

tstuefe commented:
> I had an initial look at this. Seems okay in principle. The naming/terminology needs some updating IMO: "trimNative" doesn't convey enough information, please use "trimNativeHeap". "trim" for logging tag is also too non-descript.
> 
> As this is initially experimental I'm not overly concerned about the impact, though I do cringe at yet-another-VM-thread.
> 
> FYI I will be away until next Thursday, but no need to wait for me for further comments.

Thank you @dholmes-ora. 

I dislike the introduction of another thread too, but the benefits are undeniably there, and we need  to keep the invocation of malloc_trim away from other code paths. That said, maybe a future glibc will offer us a ways to influence trim to make its runtime predictable. Which would remove the need of an own thread.

tstuefe commented on src/hotspot/share/runtime/trimNative.cpp:
```diff
@@ -0,0 +1,246 @@
+    assert(os::can_trim_native_heap(), "Unexpected");
+    os::size_change_t sc;
+    Ticks start = Ticks::now();
+    log_debug(trim)("Trim native heap started...");
```
I'll revert to using elapsedTime for all timing needs here.

TraceTime seems too much hassle. I want the time combined with other output in one line, and I don't want to provide a TraceTimerLogPrintFunc.

tstuefe commented:
Thanks @lkorinth @shipilev @robehn for the reviews.

Next version:
- renamed TrimNative namespace to NativeHeapTrimmer, the log tag to "trimnh", files, include guards etc
- Reworked the trimmer thread to:
   - uniformly use elapsedTime
   - Use atomics for the trim count
   - get rid of run_inner
   - made suspend count 16bit
   - tightened code around the trim loop
- added minimal gtest
- reworked jtreg test to use whitebox (which required a new WB method for pre-touching memory) and reshaped the test according to Alekseys suggestions.
- and lots of other smaller stuff.

Just went through the remarks again and I hope I got everything.

ashu-mehra commented:
> (just noticed the patch adds +666 lines, bad sign, I should add another line somewhere).

It also deletes 2 lines so that makes it 664 :wink:

tstuefe commented:
> > (just noticed the patch adds +666 lines, bad sign, I should add another line somewhere).
> 
> It also deletes 2 lines so that makes it 664 wink

We are good then :-)

shipilev commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,219 @@
+            const int64_t wait_ms = MAX2(1.0, to_ms(next_trim_time - tnow));
+            ml.wait(wait_ms);
+          } else if (at_or_nearing_safepoint()) {
+            ml.wait(safepoint_poll_ms);
```
OK, so here is a little problem. Suppose I want to run trims very often, like every 10ms. This loop would stall for 250ms when safepoint is detected, which throws off this guarantee. Can we instead go and sleep for `TrimNativeHeapInterval`? AFAICs, this plays nicely with heuristic guidance (short intervals -> more interference), and it would best-effort stall for twice the interval when safepoint interjects.

tstuefe replied:
But then we have a problem for larger trim intervals. Loosing one or multiple trim attempts because a safepoint happened to happen hurts if the interval is e.g. 5 minutes.

We could either wait for `MIN2(TrimNativeHeapInterval, safepoint_poll_ms)`.

Or, at the cost of one Mutex grab per safepoint, I could do a `notify_all()` at the end of a safepoint.

tstuefe replied:
It was arbitrarily chosen to have a higher chance of "slipping" in between safepoints for larger trim intervals, but not be too small to save CPU. Admittedly I thought about this less time that this sentence takes writing.

shipilev reviewed (commented):
This is a nice progress! Another read follow.

Bikeshedding: I think we also need to decide how we call this thing (and related symbols): "native heap trim" or "trim native heap". AFAICT, from there, the log tag name follows, the options name follow, the VM symbol names follow.

tstuefe commented:
New Version:

Mostly all feedback for @shipilev. Other than that:
- slightly changed the way we log when trimming (one line only)
- only calculated RSS reduction if we actually log, saves an access to procfs, which may matter for small intervals
- slight test tweaks

Re-executed the tests for release, fastdebug, fastdebug x86.

ashu-mehra commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,229 @@
+      if (logging_enabled) {
+        double t2 = now();
+        if (sc.after != SIZE_MAX) {
+          const size_t delta = sc.after < sc.before ? (sc.before - sc.after) : (sc.after - sc.before);
```
@tstuefe under what situations can `sc.after` be more than `sc.before` after trimming? Is it to handle the case where memory allocations happened in-between the malloc_trim() and the calls to get process memory?

tstuefe replied:
Yes. The numbers we print out are RSS; there is no way to get the exact figure for how much memory has been released by the glibc (not even itself knows). RSS is influenced by many factors, and its a rough outline only.

shipilev commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,229 @@
+            const int64_t wait_ms = MAX2(1.0, to_ms(next_trim_time - tnow));
+            ml.wait(wait_ms);
+          } else if (at_or_nearing_safepoint()) {
+            const int64_t wait_ms = MIN2((int64_t)TrimNativeHeapInterval, safepoint_poll_ms);
+            ml.wait(safepoint_poll_ms);
```
`MIN2<int64_t>(..., ...)` might work better without a cast?
Also, let's actually use `wait_ms` here :)

shipilev commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,229 @@
+        } while (at_or_nearing_safepoint() || is_suspended() || next_trim_time > tnow);
+      } // Lock scope
+
+      // 2 - Trim outside of lock protection.
```
There is no `1 -` to match this `2 -` to.

shipilev commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,229 @@
+
+          tnow = now();
+
+        } while (at_or_nearing_safepoint() || is_suspended() || next_trim_time > tnow);
```
Maybe invert this to pre-condition while?

shipilev reviewed (commented):
Hopefully a final read. I think there are minor things left, see comments. Alternatively, apply this patch over your current PR, which contains fixes for my comments, and then some polishing: [trimnative-shipilev-1.patch](https://github.com/openjdk/jdk/files/12025768/trimnative-shipilev-1.patch)

shipilev commented:
...also, Windows builds are failing.

shipilev commented:
One more thing, as I play with it: the GC logging does not have a comma before timestamp, see:

```
[1.210s][info][trimnative] Trim native heap (1): RSS+Swap: 1192M->1191M (-1552K), 0.353ms
[1.528s][info][gc        ] GC(0) Pause Young (Normal) (G1 Evacuation Pause) 91M->78M(1024M) 73.040ms
```

tstuefe commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,229 @@
+  uint16_t _suspend_count;
+
+  // Statistics
+  unsigned _num_trims_performed;
```
No, since `_num_trims_performed` is the number of trims performed during the lifetime of the JVM. It should probably bumped to 64-bit, now that we have millisecond intervals.

tstuefe replied:
Ah, I see you meant uint64_t, at least your patch says so.

tstuefe commented:
Thanks @shipilev!

We are closing in. 

Applied your patch, fixed Windows.

tstuefe commented:
Oh wait, I forgot your "RSS+Swap" trim message suggestion. One sec...

shipilev commented:
Yes, the only thing left is to bikeshed the logging statements a little.

tstuefe commented:
> Yes, the only thing left is to bikeshed the logging statements a little.

Okay done.

shipilev commented:
Want to replace "Native heap trimmer" with "Periodic native heap trimmer" too? Would be clear that we are suspending only the periodic one. The DCmd command would still be accepted and acted upon. Thinking about it, maybe we should do a follow-up PR and just forward that request to this thread? If so, we don't need to rename it to "Periodic".

shipilev commented:
Realized that in production, we would like to see why trimmer might be late.
I think this would look even better: [trimnative-shipilev-2.patch](https://github.com/openjdk/jdk/files/12043977/trimnative-shipilev-2.patch)

dholmes-ora commented on test/hotspot/gtest/runtime/test_trim_native.cpp:
```diff
@@ -0,0 +1,48 @@
+    return;
+  }
+
+  // Try recursive pausing
+  {
+    NativeHeapTrimmer::SuspendMark sm1("Test1");
+    {
+      NativeHeapTrimmer::SuspendMark sm2("Test2");
+      {
+        NativeHeapTrimmer::SuspendMark sm3("Test3");
```
What is this actually testing? Everything could be a no-op.

tstuefe replied:
Whether it blows up or not with internal asserts that test suspend count overflow.

But you are right; this raises another issue. One is that this is never tested since we need a gtest variant running with native trimming enabled. Another is that we need a printout of the trim state, at least for VMError::report, and that could be used here too to check the expected suspend count.

dholmes-ora reviewed (commented):
How have you determined all of the places that need `NativeHeapTrimmer::SuspendMark`?

Couple of minor comments below but otherwise this seems good. I'll take it for a spin through our CI.

Thanks

tstuefe commented:
> Realized that in production, we would like to see why trimmer might be late. I think this would look even better: [trimnative-shipilev-2.patch](https://github.com/openjdk/jdk/files/12043977/trimnative-shipilev-2.patch)

I thought about this too, but you don't really want to know if it was suspended for every wait interval, but for every trim interval. In other words, you want to know how many trims had been moved up because a safepoint had been happening, and how many trims had been skipped due to pause. Getting these infos is harder than just increasing a counter. Is it worth the added complexity?

tstuefe commented:
> Want to replace "Native heap trimmer" with "Periodic native heap trimmer" too? Would be clear that we are suspending only the periodic one. The DCmd command would still be accepted and acted upon. Thinking about it, maybe we should do a follow-up PR and just forward that request to this thread? If so, we don't need to rename it to "Periodic".

I had initially considered it and then avoided since it would add too much unnecessary complexity. E.g., you then would have to start the native trimmer thread even if periodic trimming is off in the off-chance that someone may want to issue a trim from outside. Or, you'd need to start the trimmer thread delayed and on-demand if someone issues a dcmd; then you need to deal with the fact that the trimmer thread may or may not be there. 

Another issue then would be whether trimming would expedite the current trim, aka restart the interval timer, or be additive to the periodic trim. 

My first solution did all that, and it was a lot more complex. I like the simplicity of this patch. Customers issuing trims concurrently via dcmd will hopefully be a very rare occurrence - all those folks that used to issue the dcmd scripted every second hopefully will just switch to the periodic trimming.

Idk. Its all solvable, its just code I guess. Maybe for a later RFE? Integrating the dcmd with periodic trimming may have one pro, that is preventing customers from shooting themselves in the foot who issue the dcmd via script and then forgot they did that :-P

tstuefe commented:
Hi David,

> How have you determined all of the places that need `NativeHeapTrimmer::SuspendMark`?
> 

We now avoid trimming if the VM is in or nearing a SafePoint, which takes care of the majority of phases where we don't want to trim. The rest of the suspensions was just based on my gut feeling: mallocs are usually spread out more, but some releases happen in bulk, and those are easily identifiable and have the SuspendMark.

When thinking this up, I considered a lot of alternatives, e.g., hooking into os::malloc/free and setting a dead mans timer each time to give the process a "cool down" after invocations - all in the assumption that mallocs are clustered. But that was too expensive - we did a lot of work to make the normal malloc paths with NMT=off very cheap - and also ineffective since we don't see malloc use from outside hotspot (JDK or third-party), and those are the main targets. 

To really catch every malloc, I even hooked into glibc malloc hooks, but that was horribly ugly and they removed the malloc hooks in recent glibc versions.

Bottom line, I think the current solution is a good compromise of complexity vs usefulness.

> Couple of minor comments below but otherwise this seems good. I'll take it for a spin through our CI.
> 

Thank you!

tstuefe commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,234 @@
+  }
+
+  // in seconds
+  static double now() { return os::elapsedTime(); }
```
I rather keep it since it keeps the caller sites a bit better readable.

dholmes-ora commented on test/hotspot/jtreg/runtime/os/TestTrimNative.java:
```diff
@@ -0,0 +1,308 @@
+                );
+                checkExpectedLogMessages(output, false, 0);
+                parseOutputAndLookForNegativeTrim(output, 0, 0);
+                output.shouldContain("Native trim not supported on this platform");
```
This needs updating - the current log output is:
```
Native heap trim is not supported on this platform
```

dholmes-ora reviewed (changes requested):
Test needs a fix for non-Linux.

shipilev commented:
> Idk. Its all solvable, its just code I guess. Maybe for a later RFE? Integrating the dcmd with periodic trimming may have one pro, that is preventing customers from shooting themselves in the foot who issue the dcmd via script and then forgot they did that :-P

Yes, I am not saying we should do it in this PR. Just thinking ahead on interaction with manual trim.

shipilev commented:
> > Realized that in production, we would like to see why trimmer might be late. I think this would look even better: [trimnative-shipilev-2.patch](https://github.com/openjdk/jdk/files/12043977/trimnative-shipilev-2.patch)
> 
> I thought about this too, but you don't really want to know if it was suspended for every wait interval, but for every trim interval. In other words, you want to know how many trims had been moved up because a safepoint had been happening, and how many trims had been skipped due to pause. 

Well, yes. Isn't that what my patch did?

tstuefe commented:
> > > Realized that in production, we would like to see why trimmer might be late. I think this would look even better: [trimnative-shipilev-2.patch](https://github.com/openjdk/jdk/files/12043977/trimnative-shipilev-2.patch)
> > 
> > 
> > I thought about this too, but you don't really want to know if it was suspended for every wait interval, but for every trim interval. In other words, you want to know how many trims had been moved up because a safepoint had been happening, and how many trims had been skipped due to pause.
> 
> Well, yes. Isn't that what my patch did?

I thought you wanted to collect overarching stats about how many trims got delayed. I see now you wanted to do something much simpler. Okay.

tstuefe commented:
Hi @dholmes-ora, @shipilev,

New version:

- added the log Aleksey suggested. Tested manually with +SafepointALot, works fine. Briefly considered making a patch out of it but did not since such a test is not reliable enough unless one runs a longer time, which I wanted to avoid.
- Discussion with David made me realize a status print function would be nice. Added said function. It is now called for hs_err file generation as well as in dcmd VM.info.
- I then used that function to beef up the gtest. I also now call the gtest (only the relevant os.trim... subsection) as a separate jtreg-controlled test.
- fixed the unsupported-platforms-case test

shipilev commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,271 @@
+        ml.notify_all(); // pause end
+      }
+    }
+    log_debug(trimnative)("Trim resumed after %s (%u suspend requests)", reason, n);
```
Can you do it like I did in https://github.com/openjdk/jdk/files/12043977/trimnative-shipilev-2.patch ?

We don't say "resumed" if it really want not resumed due to non-zero suspend count.

shipilev commented on src/hotspot/share/runtime/trimNativeHeap.cpp:
```diff
@@ -0,0 +1,271 @@
+        }
+      }
+
+      log_trace(trimnative)("Times %u suspended, %u timed, %u safepoint",
```
`Times: `, I think.

shipilev reviewed (approved):
I am okay with this, with only minor nits left.

tstuefe commented:
Thanks a lot @shipilev!

Fixed the last nits, and fixed an issue where the "this platform does not support..." text was not displayed with warning level and hence not visible without Xlog. Plus, test for that.

shipilev reviewed (approved):
Still okay with it.

tstuefe commented:
@dholmes-ora Are you okay with this final version? Did your CI report any problems?

dholmes-ora reviewed (approved):
Nothing further from me.

Thanks

tstuefe commented:
I'll wait for any last remarks until tomorrow morning (CET). Barring any objections, I'll push.

tstuefe commented:
Many thanks @robehn, @shipilev and @dholmes-ora !

Feels good to have this finally in.

/integrate

dholmes-ora commented:
@tstuefe unfortunately the test is failing intermittently in our CI. I will file a bug and assign it to you. Likely we will need a quick ProblemListing though. Thanks
